File: tests/workflow/consumer/rust/src/test/test_request_groupa_structb.rs
use super::{controller, samples, stat, ClientError, Consumer, StatEvent};
use tokio::sync::mpsc::UnboundedSender;

pub async fn execute(
    consumer: &mut Consumer<ClientError>,
    tx_stat: &UnboundedSender<StatEvent>,
) -> Result<(), String> {
    let mut struct_a = false;
    let mut struct_ac = false;
    let mut struct_b = false;
    while !struct_a || !struct_b || !struct_ac {
        let response = consumer
            .groupa_structb(samples::group_a::struct_b::get())
            .await
            .map_err(|e| e.to_string())?;
        tx_stat.send(StatEvent::Inc(stat::Alias::TestRequestGroupAStructB));
        match response {
            controller::GroupAStructBResponse::GroupBStructA(res) => {
                if !samples::group_b::struct_a::equal(res.clone()) {
                    return Err(format!("Invalid data received: {:?}", res));
                }
                struct_a = true;
            }
            controller::GroupAStructBResponse::GroupBGroupCStructA(res) => {
                if !samples::group_b::group_c::struct_a::equal(res.clone()) {
                    return Err(format!("Invalid data received: {:?}", res));
                }
                struct_ac = true;
            }
            controller::GroupAStructBResponse::Err(res) => {
                if !samples::group_a::struct_b::equal(res.clone()) {
                    return Err(format!("Invalid data received: {:?}", res));
                }
                struct_b = true;
            }
        };
    }
    Ok(())
}

File: tests/workflow/consumer/typescript/src/test/test_request_groupa_structb.ts
// tslint:disable: variable-name
import { Protocol, GroupAStructB } from "../consumer/index";
import { Consumer } from "../consumer";

export function test(consumer: Consumer): Promise<void> {
	return new Promise((resolve, reject) => {
		let case_GroupBStructA = false;
		let case_GroupBGroupCStructA = false;
		let case_GroupAStructB = false;
		function send(): Promise<void> {
			return new Promise((res, rej) => {
				const request = new GroupAStructB(
					Protocol.GroupA.StructB.defaults(),
					consumer
				);
				request.send().then((response) => {
					if (response instanceof Protocol.GroupA.StructB) {
						if (!case_GroupAStructB) {
							case_GroupAStructB = true;
							res(undefined);
						} else {
							rej(
								new Error(
									`Multiple times get: Protocol.GroupA.StructB`
								)
							);
						}
					} else if (response instanceof Protocol.GroupB.StructA) {
						if (!case_GroupBStructA) {
							case_GroupBStructA = true;
							res(undefined);
						} else {
							rej(
								new Error(
									`Multiple times get: Protocol.GroupB.StructA`
								)
							);
						}
					} else if (
						response instanceof Protocol.GroupB.GroupC.StructA
					) {
						if (!case_GroupBGroupCStructA) {
							case_GroupBGroupCStructA = true;
							res(undefined);
						} else {
							rej(
								new Error(
									`Multiple times get: Protocol.GroupB.GroupC.StructA`
								)
							);
						}
					}
				});
			});
		}
		function trigger() {
			send()
				.then(() => {
					if (
						case_GroupAStructB &&
						case_GroupBStructA &&
						case_GroupBGroupCStructA
					) {
						resolve(undefined);
					} else {
						trigger();
					}
				})
				.catch(reject);
		}
		trigger();
	});
}
